src/main/java/com/rabbitmq/client/observation/micrometer/MicrometerObservationCollector.java

```diff
@@ -0,0 +1,212 @@
+  }
+
+  @Override
+  public void publish(
+      PublishCall call,
+      AMQP.Basic.Publish publish,
+      AMQP.BasicProperties properties,
+      byte[] body,
+      ConnectionInfo connectionInfo)
+      throws IOException {
+    Map<String, Object> headers;
+    if (properties.getHeaders() == null) {
+      headers = new HashMap<>();
+    } else {
+      headers = new HashMap<>(properties.getHeaders());
+    }
+    PublishContext micrometerPublishContext =
+        new PublishContext(
+            publish.getExchange(),
+            publish.getRoutingKey(),
+            headers,
+            body == null ? 0 : body.length,
+            connectionInfo);
+    AMQP.BasicProperties.Builder builder = properties.builder();
+    builder.headers(headers);
+    Observation observation =
+        RabbitMqObservationDocumentation.PUBLISH_OBSERVATION.observation(
+            this.customPublishConvention,
+            this.defaultPublishConvention,
+            () -> micrometerPublishContext,
+            registry);
+    observation.start();
+    try {
+      call.publish(builder.build());
+    } catch (IOException | AlreadyClosedException e) {
+      observation.error(e);
+      throw e;
+    } finally {
+      observation.stop();
+    }
+  }
+
+  @Override
+  public Consumer basicConsume(String queue, String consumerTag, Consumer consumer) {
+    return new ObservationConsumer(
+        queue,
+        consumer,
+        this.registry,
+        this.customProcessConvention,
+        this.defaultProcessConvention);
+  }
+
+  @Override
+  public GetResponse basicGet(BasicGetCall call, String queue) {
+    return Observation.createNotStarted("rabbitmq.receive", registry)
+        .highCardinalityKeyValues(
+            KeyValues.of(
+                RabbitMqObservationDocumentation.LowCardinalityTags.MESSAGING_OPERATION.withValue(
+                    "receive"),
+                RabbitMqObservationDocumentation.LowCardinalityTags.MESSAGING_SYSTEM.withValue(
+                    "rabbitmq")))
+        .observe(
+            () -> {
+              GetResponse response = call.get();
+              if (response != null) {
+                Map<String, Object> headers;
+                if (response.getProps() == null || response.getProps().getHeaders() == null) {
+                  headers = Collections.emptyMap();
+                } else {
+                  headers = response.getProps().getHeaders();
+                }
+                DeliverContext context =
+                    new DeliverContext(
+                        response.getEnvelope().getExchange(),
+                        response.getEnvelope().getRoutingKey(),
+                        queue,
+                        headers,
+                        response.getBody() == null ? 0 : response.getBody().length);
+                Observation observation =
+                    RabbitMqObservationDocumentation.RECEIVE_OBSERVATION.observation(
```

Comment:
Is its really intentional to start another observation over here?
You already got that `Observation.createNotStarted("rabbitmq.receive")` and its `observe()`.
Why do we need this child one, please?

Reply:
So we have 2 observations here.

- `rabbitmq.receive` is the observation that is responsible for polling. Some trace (e.g. with id X) might have been initiated earlier and suddenly someone calls `basicGet` the `rabbitmq.receive` observation will warp an operation of polling a message from the broker
- `RECEIVE_OBSERVATION` is the `ReceiverContext` based operation that will continue a trace that got propagated through the messaging headers. That means that in the RabbitMq message there could be another trace (e.g. with id Y)

Let's look at it in the following way:

- Application A initiates trace with id X -> calls some logic -> polls for the message from RabbitMq (rabbitmq.receive) <THE TRACE X FINISHES HERE>
- Application B initiates trace with id Y -> calls some logic -> sends a message to RabbitMQ with tracing header containing Y -> Application A takes the message from the queue and then continues trace id Y (RECEIVE_OBSERVATION)

Since `basicGet` is a synchronous operation we were thinking of giving the user a switch to leave the `RECEIVE_OBSERVATION` started, and a scope would be opened and then it's up to the user to stop it and stop the scope. That way they will be able to continue the trace id `Y`

Reply:
But you do that manually here just like this a bit bellow:
```
                observation.start();
                observation.stop();
```
So, yeah, you connect this `RECEIVE_OBSERVATION` to the propagated one, but it is closed immediately. 
Whatever end-user would do with the propagated trace over header afterward would be just another, parallel child span, not connected to this one. Plus looks this one is going to be a child of that `rabbitmq.receive` in the beginning of this method. No? Is the parent overridden from the value from headers?

I may guess that it is OK to have a simple span connected to that propagated trace just to fulfill the the whole picture of services interaction, but is it really an intention? Wouldn't it be better to leave such a propagation up to end-user with respective API?

Reply:
> So, yeah, you connect this RECEIVE_OBSERVATION to the propagated one, but it is closed immediately.
Whatever end-user would do with the propagated trace over header afterward would be just another, parallel child span, not connected to this one. 

As I said this code 

```java
 observation.start();
                observation.stop();
```

would have to change to sth like

```java
observation.start();
if (someOptionIsTurnedOn) {
  observation.openScope();
} else {
  observation.stop();
}
```

and then the user will have to do sth like this after some listener code gets executed

```java

Scope scope = observationRegistry.getCurrentObservationScope();
if (scope != null) {
  scope.close();
  scope.getCurrentObservation().stop();
}
```

> Plus looks this one is going to be a child of that rabbitmq.receive in the beginning of this method. No? Is the parent overridden from the value from headers?

Yes, the child will be overridden. So if you started trace X then due to the fact that the message contains X headers, the trace id will continue.

> I may guess that it is OK to have a simple span connected to that propagated trace just to fulfill the the whole picture of services interaction, but is it really an intention? Wouldn't it be better to leave such a propagation up to end-user with respective API?

I don't follow, what do you mean exactly?